src/c/common.c

```diff
@@ -16,141 +16,337 @@
+        size_t type_size)
 {
-    ucdr_reset_buffer_offset(ub, 0u);
+    bool rv = false;
+    size_t alignment = ucdr_alignment(us->offset, type_size);
+
+    if (alignment <= ucdr_remaining_size(us))
+    {
+        rv = true;
+
+        do
+        {
+            size_t buffer_avsize = us->buffer_info.size - (us->offset - us->buffer_info.origin);
+            if (alignment < buffer_avsize)
+            {
+                us->offset += alignment;
+                us->iterator += alignment;
+                alignment = 0;
+            }
+            else
+            {
+                us->offset += buffer_avsize;
+                ucdr_next_buffer_info(&us->buffer_info);
+                us->iterator = us->buffer_info.data;
+                alignment -= buffer_avsize;
+            }
+        } while(alignment > 0);
+    }
+
+    return rv;
 }
 
-void ucdr_reset_buffer_offset(ucdrBuffer* ub, size_t offset)
+// -------------------------------------------------------------------
+//                       INTERNAL UTIL IMPLEMENTATION
+// -------------------------------------------------------------------
+bool ucdr_init_buffer_info(
+        ucdrBufferInfo* binfo,
+        size_t origin,
+        uint8_t* data,
+        size_t size)
 {
-    ub->iterator = ub->init + offset;
-    ub->last_data_size = 0u;
-    ub->error = false;
+    bool rv = false;
+    if (UCDR_BUFFER_INFO_SIZE < size)
+    {
+        binfo->origin = origin;
+        binfo->data = data;
+        binfo->size = size - UCDR_BUFFER_INFO_SIZE;
+        binfo->next = NULL;
+        binfo->prev = NULL;
+        memcpy(binfo->data + binfo->size, binfo, UCDR_BUFFER_INFO_SIZE);
+        rv = true;
+    }
+    return rv;
 }
 
-void ucdr_align_to(ucdrBuffer* ub, size_t size)
+bool ucdr_init_list(
+        ucdrStream* us)
 {
-    ub->iterator += ucdr_buffer_alignment(ub, size);
-    if(ub->iterator > ub->final)
+    if ((NULL != us->buffer_info.next) || (NULL != us->buffer_info.prev))
     {
-        ub->iterator = ub->final;
+        return true;
     }
 
-    ub->last_data_size = (uint8_t)size;
+    bool rv = false;
+    if (ucdr_init_buffer_info(&us->buffer_info, us->origin, us->iterator, us->size))
+    {
+        us->size = us->buffer_info.size;
+        rv = true;
+    }
+    return rv;
+}
+
+void ucdr_update_buffer_info(
+        ucdrBufferInfo* binfo)
+{
+    memcpy(binfo->data + binfo->size, binfo, UCDR_BUFFER_INFO_SIZE);
 }
 
-size_t ucdr_alignment(size_t current_alignment, size_t data_size)
+void ucdr_link_buffers(
+        ucdrBufferInfo* prev_binfo,
+        ucdrBufferInfo* next_binfo)
 {
-    return ((data_size - (current_alignment % data_size)) & (data_size - 1));
+    prev_binfo->next = next_binfo->data + next_binfo->size;
+    next_binfo->prev = prev_binfo->data + prev_binfo->size;
+    ucdr_update_buffer_info(prev_binfo);
+    ucdr_update_buffer_info(next_binfo);
 }
 
-size_t ucdr_buffer_alignment(const ucdrBuffer* ub, size_t data_size)
+bool ucdr_next_buffer_info(
```

Comment:
This method is used to iterate over the `ucdrBufferInfo`s (mostly only reading the `cudrBufferInfo`). Why you use a copy instead of return the pointer to the `ucdrBufferInfo`?

Reply:
Because I check if there is `next` buffer available.